### PR TITLE
User and UserProfile data models

### LINKED ISSUE #50: Prototype database schema
_Outcome_: Update the current data modeling PR for review, and merge initial data models to support login.gov flow. 

- [ ] #92 
- [ ] #119 (not high-priority)

### PR DESCRIPTION
**Pull request by neilmb**
# Add custom User and related UserProfile model

## 🗣 Description ##

This adds data models for a User and a one-to-one related UserProfile. The User model is un-customized from Django's default, but Django recommends using a custom model because we can't change it after our very first set of migrations.

There is also a very, very basic page at `/edit_profile/` that allows users to edit just one field of the UserProfile model, their "Display Name". It isn't intended that this page will necessarily be part of the Get.gov product, that would take more product discussion and design work. Rather it exercises the patterns for making custom forms in Django that we are going to be using over and over again on this project.

Screen Shot 2022-09-23 at 13 47 22

## 💭 Motivation and context ##

Closes #50 for the very first stages of making a database schema.

This PR will require developers to re-run migrations, possibly on a clean database inside of their Docker container.  The command `docker compose run --rm db psql -d postgres://user@db/app` can be used to get access to a SQL shell in the local db where `DROP DATABASE` and `CREATE DATABASE` can be used.

### CODE REVIEW

File: src/registrar/admin.py
Comment (SSPJ) [inline src/registrar/admin.py:11]: Docstring.

File: src/registrar/apps.py
Comment (SSPJ) [inline src/registrar/apps.py:17]: Kudos for explaining the `#noqa`. ⭐

File: src/registrar/config/settings.py
Comment (SSPJ) [inline src/registrar/config/settings.py:58]: Put this in the section with the other database settings?

File: src/registrar/forms.py
Comment (SSPJ) [inline src/registrar/forms.py:5]: Docstring.

It can feel redundant to put class documentation on a class which is so obviously self-explanatory, but future maintainers may not be as familiar with Django as you are. It is a kindness to them to include a brief but clear explanation of why a piece of code exists.

File: src/registrar/migrations/0003_alter_userprofile_street1.py
Comment (SSPJ) [inline src/registrar/migrations/0003_alter_userprofile_street1.py:3]: As this is the first PR to include migrations, could I trouble you to squash them? (Just delete and regenerate.) Unless you believe they are logically better represented as three files. That's fine as well.

File: src/registrar/models/models.py
Comment (SSPJ) [inline src/registrar/models/models.py:12]: What happens if `self.userprofile` is null?

File: src/registrar/models/models.py
Comment (SSPJ) [inline src/registrar/models/models.py:74]: Is the `on_delete` attribute uni-directional for a `OneToOneField`? (Deleting UserProfile deletes User, but deleting User does _not_ delete UserProfile.) I assume so, but want to confirm.

File: src/registrar/models/models.py
Comment (SSPJ) [inline src/registrar/models/models.py:81]: What will happen if `self.user` is null?

File: src/registrar/signals.py
Comment (SSPJ) [inline src/registrar/signals.py:9]: Docstring.

File: src/registrar/signals.py
Comment (SSPJ) [inline src/registrar/signals.py:15]: Docstring.

File: src/registrar/signals.py
Comment (SSPJ) [inline src/registrar/signals.py:17]: Suggest checking for the existence of `userprofile` and creating it if it does not exist.

File: src/registrar/tests/test_views.py
Comment (SSPJ) [inline src/registrar/tests/test_views.py:17]: Thank you for including tests! 🎸

File: src/registrar/views/profile.py
Comment (SSPJ) [inline src/registrar/views/profile.py:10]: Docstring.

File: src/registrar/models/models.py
Comment (neilmb) [reply src/registrar/models/models.py:74]: It is uni-directional, but it goes the opposite of the direction that you specify. Even though we are specifying it here on the `UserProfile` model, `on_delete` describes what should happen when we delete the matching `User`. 

In this case, a `User` who is deleted no longer needs a `UserProfile` so we want to also remove the `UserProfile` at that time which is what `CASCADE` indicates.

https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.ForeignKey.on_delete for the formal documentation which just refers to SQL's behavior and https://stackoverflow.com/a/38389488 for a better explanation.

File: src/registrar/models/models.py
Comment (neilmb) [reply src/registrar/models/models.py:81]: In this case, `self.user` should never be null because of the foreign key constraint and the cascading deletes. But for safety, I can change this to a guarded `getattr` as well.

File: src/registrar/models/models.py
Comment (SSPJ) [reply src/registrar/models/models.py:74]: Ack, thank you for the docs link! I do tend to get confused by foreign key constraints since the key is so often held in the table which is opposite to the "natural" relationship (e.g. a car has wheels, but a Wheel has the key to the Car...)

I have a [draft ADR](https://docs.google.com/document/d/1ea04AvWeXwxelxmhD2rqMfcRBnKdnxvam6EyA8ehe7I/edit#bookmark=id.po74ml55c5lk) for the business logic decisions that are going into the design of the database models (e.g. "a User who is deleted no longer needs a UserProfile"), but it is kinda messy and has met with some objections from the team. Would you be good enough to start a new one in that document and include your thinking behind the models on this PR?

**Review by SSPJ — APPROVED** [review]
LGTM 👍  Nice work, Neil!

I was able to test this on dev.

As we discussed in chat, there is some concern about whether the signal handler is working as expected. We can test this again once it and #139 have merged and make any necessary adjustments.

Feel free to add you ADR to this PR or open a follow up one for it.

**Pull request merged**
